dds/DCPS/XTypes/IdlScanner.h

```diff
@@ -0,0 +1,890 @@
+    } else {
+      return true;
+    }
+  }
+
+  IdlToken scan_numeric_before_decimal()
+  {
+    if (scan_numeric_integer_digit() && scan_numeric_integer_optional_digits()) {
+      if (scanner_.eoi()) {
+        return IdlToken::make_integer(signed_, integer_value_);
+      }
+
+      const char c = scanner_.peek();
+      if (c == '.') {
+        if (scanner_.match(c) && scan_numeric_after_decimal(true)) {
+          return make_float();
+        } else {
+          return IdlToken::make_error();
+        }
+      } else if (c == 'e' || c == 'E') {
+        if (scanner_.match(c) && scan_numeric_exponent()) {
+          return make_float();
+        } else {
+          return IdlToken::make_error();
+        }
+      } else {
+        return IdlToken::make_integer(signed_, integer_value_);
+      }
+    } else {
+      return IdlToken::make_error();
+    }
+  }
+
+  bool scan_numeric_integer_digit()
+  {
+    if (scanner_.eoi()) {
+      return false;
+    }
+
+    const char c = scanner_.peek();
+    if (c >= '0' && c <= '9') {
+      return scanner_.match(c) && scale_integer(10, c - '0');
+    } else {
+      return false;
+    }
+  }
+
+  bool scan_numeric_integer_optional_digits()
+  {
+    if (scanner_.eoi()) {
+      return true;
+    }
+
+    const char c = scanner_.peek();
+    if (c >= '0' && c <= '9') {
+      return scanner_.match(c) && scale_integer(10, c - '0') && scan_numeric_integer_optional_digits();
+    } else {
+      return true;
+    }
+  }
+
+  IdlToken make_float()
+  {
+    while (digits_after_decimal_) {
+      if (exponent_signed_) {
+        ++exponent_value_;
+        if (exponent_value_ == 0) {
+          return IdlToken::make_error();
+        }
+      } else {
+        if (exponent_value_ > 0) {
+          --exponent_value_;
+        } else {
+          exponent_signed_ = true;
+          exponent_value_ = 1;
+        }
+      }
+      --digits_after_decimal_;
+    }
+    return IdlToken::make_float(signed_, integer_value_, exponent_signed_, exponent_value_);
+  }
+
+  bool scale_integer(ACE_UINT64 base, ACE_UINT64 value)
+  {
+    const ACE_UINT64 next = integer_value_ * base + value;
+    if (integer_value_ && next < integer_value_) {
+      return false;
+    }
+    integer_value_ = next;
+    return true;
+  }
+
+  bool scale_exponent(ACE_UINT64 value)
+  {
+    const ACE_UINT64 next = exponent_value_ * 10 + value;
+    if (exponent_value_ && next < exponent_value_) {
+      return false;
+    }
+    exponent_value_ = next;
+    return true;
+  }
+
+  CharacterScanner& scanner_;
+  union {
+    bool boolean_literal_;
+    char character_literal_;
+    struct {
+      bool signed_;
+      ACE_UINT64 integer_value_;
+      ACE_UINT64 digits_after_decimal_;
+      bool exponent_signed_;
+      ACE_UINT64 exponent_value_; // TODO:  Handle overflow.
+    };
+  };
+  std::string string_literal_;
+  std::string identifier_;
```

Comment:
In another comment on `IdlToken`, my thought was that union may be used to make it clear that an instance of `IdlToken` (and `IdlScanner` if I understand it right) can hold only one of those values at a time. But looks like `string` can't be included in a union. I'm not sure what an alternative would be.

Also, there are some similarities between the data members of `IdlToken` and `IdlScanner`. Could they share a common data type for that?